File: LC0589_N_ary_Tree_Preorder/n_ary_tree_preorder.cpp
#include <vector>
using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};


class Solution {
private:
    vector<int> result;
public:
    // constructor
    Solution()
    {
        this->result = vector<int>();
    }

    void rPreorder(Node* root)
    {
        if (root == nullptr)
        {
            return;
        }
        this->result.push_back(root->val);
        for (Node* child : root->children)
        {
            this->rPreorder(child);
        }
    }

    vector<int> preorder(Node* root) {
        if (root == nullptr)
        {
            return vector<int>();
        }
        this->rPreorder(root);
        return this->result;
    }
};

File: LC0589_N_ary_Tree_Preorder/n_ary_tree_preorder.java
import java.util.List;
import java.util.ArrayList;

// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};


class Solution {
    private List<Integer> result; 
    public Solution() {
        this.result = new ArrayList<Integer>(); // stores the answer
    }

    public void rPreorder(Node root)
    {
        if (root == null)
        {
            return;
        }
        this.result.add(root.val);
        for (Node child : root.children)
        {
            this.rPreorder(child);
        }
    }

    public List<Integer> preorder(Node root) {
        if (root == null)
        {
            return new ArrayList<Integer>();
        }
        this.rPreorder(root);
        return this.result;
    }
}